Notify about recent pushes if a branch has no PR yet

kolaente opened the pull request:
This PR adds a nice little message on the dashboard if the current user recently pushed to a branch which has no PR open yet, with a link to do so. If the user recently pushed to a fork, it will show a message as well, but the link will go to create a PR with the upstream repository instead.

This PR explicitly does not show a link to existing PRs after pushes to a branch to keep it simple.

### Screenshots

![image](https://user-images.githubusercontent.com/13721712/229354954-6dfc3a13-8d04-44d2-bbe9-8cdd7f5e7c6f.png)

![image](https://user-images.githubusercontent.com/13721712/229354980-48db3f81-ad34-4d03-a5f7-2dc09b588725.png)

![image](https://user-images.githubusercontent.com/13721712/229498150-70b8c2f8-f51f-4c7c-a0cc-2decc91af687.png)

![image](https://user-images.githubusercontent.com/13721712/229498180-2f75cd76-81ce-47aa-80d0-17c38cab789b.png)

![image](https://user-images.githubusercontent.com/13721712/229498210-a5191947-f816-42f4-a086-201d61f9c8c3.png)


### TODO

* [x] Make sure it is working for both pushes to a new branch on the same repo and on a fork
* [x] Actually show the message on the dashboard
* [x] Hide pushes to the default branch since it does not make sense to create PRs for that _unless_ the push was made to a forked repo
* [x] Include the time when the push happened in the message
* [x] Add a screenshot here

Resolves #311, resolves #13196, resolves #23743

CirnoT commented:
I'm worried it might be too cluttered when more than few message are displayed (especially with rather long time limit). How about moving it to repo page similar to how GH does it?

lunny commented:
How about display it in the repository home page but not user's dashboard page. Once the user push to serval repositories, which one should we display?

kolaente commented:
> I'm worried it might be too cluttered when more than few message are displayed (especially with rather long time limit). How about moving it to repo page similar to how GH does it?

@CirnoT I think this could be addressed with either a shorter time limit or limiting the max messages to 2 or 4 (or both)

I think github shows it on both the repo page and the dashboard. IIRC gitlab does that as well.

lunny commented:
Fix #23743

kolaente commented:
Would there be interest in picking this PR up again?

lunny commented:
Maybe it should be on top of the repository home page but not dashboard?

delvh commented:
The interest is definitely there, but I wonder if it wouldn't make more sense to create a new branch, one to resolve all conflicts that happened in the mean time, and two to give it the attention boost of a new PR?

kolaente commented:
@delvh Probably. Should I do that or wants someone else pick it up?

delvh commented on templates/user/recently_pushed_branches.tmpl:
```diff
@@ -0,0 +1,8 @@
+{{range .RecentlyPushedBranches}}
+    <div class="ui blue message">
+        {{$.i18n.Tr "repo.pulls.recently_pushed_to_branches" .Repo.Owner.Name .Repo.LowerName .RefName | Safe}}
+    	<a class="ui compact blue button" href="/{{.Repo.Owner.Name}}/{{.Repo.LowerName}}/compare/{{.Repo.DefaultBranch | EscapePound}}...{{if .Repo.IsFork}}{{.Repo.Owner.Name}}:{{end}}{{.RefName | EscapePound}}">
```
Yes

kolaente replied:
Looks like there's a function `CompareLink` which does exactly this.

kolaente commented on models/action.go:
```diff
@@ -395,3 +395,67 @@ func activityQueryCondition(opts GetFeedsOptions) (builder.Cond, error) {
+			),
+			builder.Gte{"action.created_unix": limit},
+		)).
+		Limit(10).
```
Done.

kolaente commented:
This is now ready for review.

yardenshoham reviewed (changes requested):
CI failure is related

yardenshoham commented:
![image](https://user-images.githubusercontent.com/20454870/229309067-79275b20-e27c-4b47-b694-b54bab895ba1.png)
I wonder if we should somehow restrict it to always display as a single line?

kolaente commented:
>  I wonder if we should somehow restrict it to always display as a single line?

I've now changed it to this:

![image](https://user-images.githubusercontent.com/13721712/229310253-e4d35737-cbb8-43fe-afbf-870aae3bdc2b.png)

yardenshoham commented:
![image](https://user-images.githubusercontent.com/20454870/229310483-ade8275a-3058-4b3c-9134-8d4eed9f3558.png)
Not sure why but the bold parts barely stand out in the dark theme, it might just be me though 🤷‍♂️

yardenshoham commented on models/recently_pushed.go:
I would greatly appreciate it if you could include some comments to help me better comprehend the underlying logic. Given my limited experience with golang/xorm, a more detailed explanation would be immensely helpful before I can confidently give my approval

kolaente replied:
Added some comments, please check back

kolaente commented:
Now the tests are failing in postgres with this error message:

>     testlogger.go:77: 2023/04/01 20:25:58 ...ers/web/user/home.go:133:Dashboard() [E] [64289356-9] GetRecentlyPushedBranches: pq: invalid input syntax for type bigint: "issue.closed_unix"

I don't know what could cause the problem here. Both types being compared are bigints. Theoretically this should not be a problem.

yardenshoham reviewed (commented):
Looks very good to me. Is there a reason to use blue instead of the usual theme? Maybe then the bold/strong parts will look better in the dark theme than they currently do.

I will approve if the build passes

lunny commented on models/recently_pushed.go:
```diff
@@ -0,0 +1,123 @@
+				builder.And(
+					builder.Eq{"pull_request.has_merged": false},
+					builder.Eq{"issue.is_closed": true},
+					builder.Gt{"action.created_unix": "issue.closed_unix"},
```
You have to use `builder.Expr("action.created_unix = issue.closed_unix")`.

kolaente replied:
Done

kolaente commented:
@yardenshoham Not really. I thought it looks good that way. I think it doesn't stand out enough if I don't use the blue style:

![image](https://user-images.githubusercontent.com/13721712/229353195-d3463486-19ff-495a-b6dc-59c76ebe68c0.png)

`primary` doesn't seem to work great either. Do you have another suggestion?

yardenshoham commented:
How about green instead of blue to match Gitea's theme? Try to use the dark theme in your setup to see it how I do

yardenshoham commented:
Maybe @silverwind or @wxiaoguang would know which CSS setup will fit here

yardenshoham reviewed (changes requested):
CI failure is related

kolaente commented:
I think just plain `green` looks pretty awful:

![image](https://user-images.githubusercontent.com/13721712/229354866-aaed6ed7-14bc-4afd-9561-9f855904eb4a.png)

yardenshoham commented:
Yeah, I guess we can stick with blue

kolaente commented:
Took another look and it seems like there's a class called `positive` which seems to work well:

![image](https://user-images.githubusercontent.com/13721712/229354954-6dfc3a13-8d04-44d2-bbe9-8cdd7f5e7c6f.png)

![image](https://user-images.githubusercontent.com/13721712/229354980-48db3f81-ad34-4d03-a5f7-2dc09b588725.png)

kolaente commented:
Tests are passing now :tada:

kolaente commented:
@delvh @lunny Can you check again?

yardenshoham commented:
@kolaente don't forget to update the PR description with the latest info and screenshots

CirnoT commented:
Still think this belongs on individual repo page.

kolaente commented:
> Still think this belongs on individual repo page.

I mean we could probably do both (as GitHub does it) but I think that should be a follow-up PR.

wxiaoguang commented on models/recently_pushed.go:
```diff
@@ -0,0 +1,123 @@
+		pushed := &RecentlyPushedBranches{
+			Repo:     repos[a.RepoID],
+			BaseRepo: repos[a.RepoID],
+			RefName:  strings.Replace(a.RefName, "refs/heads/", "", 1),
```
If the prefix is stripped, it's not a "RefName" anymore. It's a "BranchName"

kolaente replied:
Good catch! Changed.

wxiaoguang commented on web_src/css/home.css:
```diff
@@ -51,3 +51,12 @@
   }
 }
+
+.ui.message.recently-pushed-branches {
+  align-items: center;
+  padding: 0;
+}
+
+.ui.message.recently-pushed-branches .button {
+  margin: 0;
+}
```
I guess there are some helper styles, like `gt-ac`, `gt-p-0` and `gt-m-0`, then we do not need to add more styles.

kolaente replied:
I thought that as well, but wasn't able to find them. Is there an overview with all these helper styles somewhere?

delvh replied:
`web_src/css/helpers.css`

wxiaoguang replied:
https://github.com/go-gitea/gitea/blob/main/web_src/css/helpers.css

kolaente replied:
Thanks! I've now changed it to use the helper classes instead.

silverwind commented:
Agree, should be on repo page, but I guess I won't block it if you want it on frontpage. Would not do both positions.

kolaente commented:
> Agree, should be on repo page, but I guess I won't block it if you want it on frontpage. Would not do both positions.

In my workflow, I'm pushing to a new branch and then just opening the root domain of my Gitea instance. If the message offering me to create a new PR from my recent changes is right there on the dashboard, it saves me a few clicks.

> I do see the risk of I push to many branches where I don't want to open a PR, it'll litter the frontpage with these.

That's why we only show the last three.

silverwind commented on models/recently_pushed.go:
```diff
@@ -0,0 +1,123 @@
+
+// GetRecentlyPushedBranches returns all actions where a user recently pushed but no PRs are created yet.
+func GetRecentlyPushedBranches(ctx context.Context, u *user.User) (recentlyPushedBranches []*RecentlyPushedBranches, err error) {
+	limit := time.Now().Add(-24 * time.Hour).Unix()
```
24 hours seems a bit long. I have a feeling this is much shorter on GitHub. Anyone knows how long it is? I would say 2-4 hours max.

silverwind replied:
As per https://stackoverflow.com/a/23129795/808699 it seems to be 2 hours.

kolaente replied:
I've made it two hours now.

If there are people who want to change this in the future, we may add a config variable to do this.

silverwind commented:
> In my workflow, I'm pushing to a new branch and then just opening the root domain of my Gitea instance. If the message offering me to create a new PR from my recent changes is right there on the dashboard, it saves me a few clicks.

With GH, I usually have the target repo open and expect the popup on either repo root, issue list or pr list. I would display it in all those three places there. I barely use the "home" page on either gitea or github.

kolaente commented:
I've added it to the repo as well:

![image](https://user-images.githubusercontent.com/13721712/229498150-70b8c2f8-f51f-4c7c-a0cc-2decc91af687.png)

![image](https://user-images.githubusercontent.com/13721712/229498180-2f75cd76-81ce-47aa-80d0-17c38cab789b.png)

![image](https://user-images.githubusercontent.com/13721712/229498210-a5191947-f816-42f4-a086-201d61f9c8c3.png)

kolaente commented on templates/user/recently_pushed_branches.tmpl:
```diff
@@ -0,0 +1,13 @@
+				{{$.locale.Tr "repo.pulls.compare_changes"}}
+			</a>
+		</div>
+	</div>
```
That makes sense. It almost works, I've made it work based on your suggestion.

silverwind commented:
Can you left-align the text? It seems off in those screenshots. Will probably test myself later today.

kolaente commented:
> Can you left-align the text? It seems off in those screenshots. Will probably test myself later today.

Sure.

yardenshoham reviewed (changes requested):
CI failure is related

lunny commented:
I created a branch on the base repository from the UI, but no hint display.

kolaente commented:
@lunny  It worked for me: 

![image](https://user-images.githubusercontent.com/13721712/229839987-123643a6-7579-4068-a21a-03d7b6ed1be1.png)

Did you only create the branch? (without changes)? Where did you expect the message?

kolaente commented:
The ci failure seems unrelated, can someone restart the build?

yardenshoham commented:
`git commit -m "Trigger build" --allow-empty`

lunny commented:
The SQL logs looks like this

```log
2023/04/07 15:40:47 ...s/recently_pushed.go:60:GetRecentlyPushedBranches() [I] [SQL] SELECT action.ref_name, action.repo_id, action.created_unix FROM "action" LEFT JOIN "pull_request" ON pull_request.head_branch = replace(action.ref_name, 'refs/heads/', '') LEFT JOIN "issue" ON pull_request.issue_id = issue.id LEFT JOIN "repository" ON action.repo_id = repository.id WHERE action.op_type=$1 AND action.act_user_id=$2 AND ((repository.default_branch != replace(action.ref_name, 'refs/heads/', '')) OR repository.is_fork=$3) AND (pull_request.id IS NULL OR (pull_request.has_merged=$4 AND issue.is_closed=$5 AND (action.created_unix > issue.closed_unix))) AND action.created_unix>=$6 GROUP BY action.ref_name, action.repo_id, action.created_unix ORDER BY "action"."created_unix" DESC LIMIT 3 [commit_repo 3 true false true 1680846047] - 13.430291ms
2023/04/07 15:40:47 ...s/recently_pushed.go:82:GetRecentlyPushedBranches() [I] [SQL] SELECT "id", "owner_id", "owner_name", "lower_name", "name", "description", "website", "original_service_type", "original_url", "default_branch", "num_watches", "num_stars", "num_forks", "num_issues", "num_closed_issues", "num_pulls", "num_closed_pulls", "num_milestones", "num_closed_milestones", "num_projects", "num_closed_projects", "num_action_runs", "num_closed_action_runs", "is_private", "is_empty", "is_archived", "is_mirror", "status", "is_fork", "fork_id", "is_template", "template_id", "size", "is_fsck_enabled", "close_issues_via_commit_in_any_branch", "topics", "trust_model", "avatar", "created_unix", "updated_unix" FROM "repository" WHERE (0=1 OR repository.id IN (SELECT repository.fork_id FROM repository WHERE 0=1)) [] - 2.824583ms
2023/04/07 15:40:47 ...s/recently_pushed.go:98:GetRecentlyPushedBranches() [I] [SQL] SELECT * FROM "user" LEFT JOIN "repository" ON "repository".owner_id = "user".id WHERE (0=1 OR repository.id IN (SELECT repository.fork_id FROM repository WHERE 0=1)) [] - 1.515667ms
```

lunny commented:
I think this should be reimplemented after #22743 is merged. It will become very easy to get recently pushed branches after that PR.

lunny reviewed (changes requested):
This PR's implementation has a big performance problem that should be resolved.

kolaente commented:
> This PR's implementation has a big performance problem that should be resolved.

What exactly do you mean? What should be changed?

> I think this should be reimplemented after https://github.com/go-gitea/gitea/pull/22743 is merged. It will become very easy to get recently pushed branches after that PR.

Does that block this PR?

wxiaoguang commented:
I think the problem could be partially resolved by this approach:

1. Only do a simple select query in action table (no JOIN). Since we only need the recent 2 hour activity record, and action table has indices, the performance should be good enough.
2. Then, use other queries to gather more details.

@lunny what do you think about this?

kolaente commented:
But isn't the performance of one query with joins pretty much always better than multiple queries? The database knows a lot more about the data in the db than we will from the outside so the optimizer will probably do a better job on this.

delvh commented:
Well, except you basically have a join over four of the largest tables in the whole Gitea code.
While it might work on small instances of up to 10 people, this is not really scalable for something of the level of `gitea.com` or even `codeberg.org`, where many millions of entries are in these tables.

CirnoT commented:
> But isn't the performance of one query with joins pretty much always better than multiple queries? The database knows a lot more about the data in the db than we will from the outside so the optimizer will probably do a better job on this.

No, not really. `JOIN` is performed before filtering (as also suggested by order in which query is written and the fact that you can filter on joined fields) so if you need to join few tables with millions of results just to fetch ~100 latest results from single table and correlate data with others, it would be much more efficient to get individual IDs via singular index than to have database do full merge of million of records into temporary table and then query that.

wxiaoguang commented:
> But isn't the performance of one query with joins pretty much always better than multiple queries? The database knows a lot more about the data in the db than we will from the outside so the optimizer will probably do a better job on this.

Theoretically yes, actually, it depends.

Databases are just "guessing" the best index, they may make mistakes. That's why MySQL has key word "USE INDEX" or even "FORCE INDEX", I have used them many times. PostgreSQL may be somewhat better, developers still could use some tricks to make it use some particular indices.

And that's why GitLab dropped other database support, only supports one database -- it's difficult to fine-tune a large dataset on different databases.

CirnoT commented:
>Databases are just "guessing" the best index, they may make mistakes. That's why MySQL has key word "USE INDEX" or even "FORCE INDEX", I have used them many times. PostgreSQL may be somewhat better, developers still could use some tricks to make it use some particular indices.

This is more of a issue with certain engines than others. For example engines such as InnoDB do not really keep stats on exact count of rows and instead estimate them - this might lead to wrong index cardinality and wrong decision being taken.

kolaente commented:
It looks like the `pull_request` table is the problem here. Explaining the query indicates it does a full table scan on that table:

```
+------+-------------+--------------+--------+----------------------------------------+-----------------------+---------+-----------------------------+------+-------------------------------------------------+
| id   | select_type | table        | type   | possible_keys                          | key                   | key_len | ref                         | rows | Extra                                           |
+------+-------------+--------------+--------+----------------------------------------+-----------------------+---------+-----------------------------+------+-------------------------------------------------+
|    1 | SIMPLE      | repository   | ALL    | PRIMARY,IDX_repository_is_fork         | NULL                  | NULL    | NULL                        | 359  | Using temporary; Using filesort                 |
|    1 | SIMPLE      | action       | ref    | IDX_action_au_r_c_u_d,IDX_action_r_u_d | IDX_action_au_r_c_u_d | 18      | const,gitea.repository.id   | 104  | Using index condition; Using where              |
|    1 | SIMPLE      | pull_request | ALL    | NULL                                   | NULL                  | NULL    | NULL                        | 5978 | Using where; Using join buffer (flat, BNL join) |
|    1 | SIMPLE      | issue        | eq_ref | PRIMARY                                | PRIMARY               | 8       | gitea.pull_request.issue_id | 1    | Using where                                     |
+------+-------------+--------------+--------+----------------------------------------+-----------------------+---------+-----------------------------+------+-------------------------------------------------+
```

And while this is indeed fast on my local dev instance, the query takes 500ms on my own Gitea instance (using mysql), which has 6456 PRs created to date (So it's not a large instance but not exactly a small one either).

My first guess was this is because there's no index on the `head_branch` column of the PR table. So I added one:

```sql
create index pull_request_action_head_branch on pull_request (head_branch);
```

Running the query again, it only took 31ms. Explaining the query again, it looks like the db now uses the index to fetch the PR details:

```
+------+-------------+--------------+--------+----------------------------------------+---------------------------------+---------+-----------------------------+------+------------------------------------+
| id   | select_type | table        | type   | possible_keys                          | key                             | key_len | ref                         | rows | Extra                              |
+------+-------------+--------------+--------+----------------------------------------+---------------------------------+---------+-----------------------------+------+------------------------------------+
|    1 | SIMPLE      | repository   | ALL    | PRIMARY,IDX_repository_is_fork         | NULL                            | NULL    | NULL                        | 359  | Using temporary; Using filesort    |
|    1 | SIMPLE      | action       | ref    | IDX_action_au_r_c_u_d,IDX_action_r_u_d | IDX_action_au_r_c_u_d           | 18      | const,gitea.repository.id   | 104  | Using index condition; Using where |
|    1 | SIMPLE      | pull_request | ref    | pull_request_action_head_branch        | pull_request_action_head_branch | 1023    | func                        | 2    | Using where                        |
|    1 | SIMPLE      | issue        | eq_ref | PRIMARY                                | PRIMARY                         | 8       | gitea.pull_request.issue_id | 1    | Using where                        |
+------+-------------+--------------+--------+----------------------------------------+---------------------------------+---------+-----------------------------+------+------------------------------------+
```

I'd love to see that experiment with the data from gitea.com.

I propose to add an index on the `pull_request` table and call it a day.

kolaente commented:
> you need to join few tables with millions of results just to fetch ~100 latest results from single table and correlate data with others, it would be much more efficient to get individual IDs via singular index than to have database do full merge of million of records into temporary table and then query that.

Except the query starts by fetching the actions from the last two hours. If that's millions of rows _for the currently logged in user_ the database is either really beafy already (because then you probably have 100k+ users and need that much power) or it's not and everything else is slow as well. I'd argue this is very unlikely and thus not a problem.

> fields that perform very costly actions such as REPLACE() - this function is executed per every row before filter is applied!

But iirc only once for each result from the `action` table? See above, that shouldn't be that many in the first place.

CirnoT commented:
> I'd love to see that experiment with the data from gitea.com.

Check this plan: https://explain.dalibo.com/plan/7fb7f63h6e0ca080

Some stats:

```
gitea=# select count(*) from pull_request;
 count 
-------
  6020
(1 row)

gitea=# select count(*) from action;
 count  
--------
 274991
(1 row)
```

The query was adjusted to fetch actions from about last 7 days to estimate how it would behave on moderately popular instance

```
gitea=# select count(*) from action where action.created_unix>=1680362902;
 count 
-------
  1183
(1 row)
```

CirnoT commented:
Here's possibly better plan where query was adjusted for a user that commits more often (bot): https://explain.dalibo.com/plan/cab6gc122759b18c

silverwind commented:
> And that's why GitLab dropped other database support, only supports one database -- it's difficult to fine-tune a large dataset on different databases.

That is my impression as well. Single database allows many optimizations and features that just are not possible with multi-database. We might consider it if we can provide migrations, but I think that will be not so easy.

CirnoT commented:
Also on unrelated note, that query returns for quite weird results:
```
gitea=# SELECT action.ref_name, action.repo_id, action.created_unix FROM "action" LEFT JOIN "pull_request" ON pull_request.head_branch = replace(action.ref_name, 'refs/heads/', '') LEFT JOIN "issue" ON pull_request.issue_id = issue.id LEFT JOIN "repository" ON action.repo_id = repository.id WHERE action.op_type=5 AND action.act_user_id=32 AND ((repository.default_branch != replace(action.ref_name, 'refs/heads/', '')) OR repository.is_fork=TRUE) AND (pull_request.id IS NULL OR (pull_request.has_merged=FALSE AND issue.is_closed=TRUE AND (action.created_unix > issue.closed_unix))) AND action.created_unix>=1680362902 GROUP BY action.ref_name, action.repo_id, action.created_unix ORDER BY "action"."created_unix" DESC;
               ref_name               | repo_id | created_unix 
--------------------------------------+---------+--------------
 refs/heads/renovate/phpstan-monorepo |       1 |   1680460273
 refs/heads/renovate/phpstan-monorepo |       1 |   1680460272
 refs/heads/renovate/phpstan-monorepo |      13 |   1680460072
 refs/heads/renovate/webpack-5.x      |       6 |   1680459707
 refs/heads/renovate/phpstan-monorepo |       1 |   1680456696
 refs/heads/renovate/phpstan-monorepo |      13 |   1680456540
 refs/heads/renovate/phpstan-monorepo |       9 |   1680456286
 refs/heads/renovate/phpstan-monorepo |       9 |   1680456285
```
These branches do not exist, Multiple PRs for them exists but they are either merged or closed (unmerged).

lunny commented:
> I think the performance problem could be partially resolved by this approach:
> 
>     1. Only do a simple select query in action table (no JOIN). Since we only need the recent 2 hour activity record, and action table has indices, the performance should be good enough.
> 
>     2. Then, use other queries to gather more details.
> 
> 
> @lunny what do you think about this?

I think we could also do some optimization at the moment but those are temporary methods. As I said, once branches have been synced into databases, this PR could be implemented very easily.  Assume we have a new table `branch` with an index `repo_id` and `updated` timestamp. We can easily get recently updated branch names of this repository or all the branches in this instance. No complex SQL, no Joins.

silverwind commented:
We could on branch push push all pushed branches into some a append-only structure based on username where values would automatically expire after the 2 hour time period (like redis TTL command).

Then, when rendering, the operation is just a lookup in that list.

The only thing I'm not sure about if auto-expiring keys can be done with SQL, or if that would require a cache like redis.

yardenshoham commented:
I think GitHub keeps it out of the initial page render. Maybe it can be done in the background with an API call and render the suggestion only when ready?

silverwind commented:
Yeah, it is definitely async on GitHub but causes ugly JS pop-in. If we can find a performant synchronous solution it would be ideal.

kolaente commented:
Thanks everyone for your investigations and queries. My goal was to use existing data structures to avoid a higher overhead for such a simple feature. Well, looks like this is not as simple as I thought.

> I think we could also do some optimization at the moment but those are temporary methods. As I said, once branches have been synced into databases, this PR could be implemented very easily. Assume we have a new table branch with an index repo_id and updated timestamp. We can easily get recently updated branch names of this repository or all the branches in this instance. No complex SQL, no Joins.

Should this PR be blocked and reworked when the other one got merged?

yardenshoham commented:
* Blocked by https://github.com/go-gitea/gitea/pull/22743

techknowlogick commented:
@yardenshoham caught up in a PR cleanup sweep, thanks for re-opening :)

wxiaoguang commented:
I think maybe this PR could be taken in 1.20 if no objection for "only querying recent N hours `actions` records by a single/simple SELECT" ?

* If the query range is limited by time and the query is simple, it doesn't seem to cause server side load problem?

silverwind commented:
Yes, and ideally only show notification on the repo page. Not on the frontpage.

lunny commented:
Since #22743 merged, I think this could be continued now.

silverwind commented:
As already mentioned, I would like to see this changed to only show on the repo page as it would be too annoying on the frontpage.

kolaente commented:
I don't really have time to implement this in the next few weeks, so if anyone wants to pick it up, feel free. Otherwise I'll take a stab at it at the end of August / early September.

6543 commented:
#25715
